sdk/mx.sdk/mx_sdk_vm_impl.py

```diff
@@ -1932,8 +1932,10 @@ def _get_option_vars():
             return ' '.join(image_config.option_vars)
 
         def _get_launcher_args():
-            if not _jlink_libraries() and _src_jdk_version >= 9:
-                return '-J--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code=jdk.internal.vm.compiler'
+            # Nashorn got removed in JDK 15, added exports for the JVMCI compiler got added
+            # with JDK-8223332 (JDK 13) and JDK-8273547 (11.0.14). Use the larger upper bound.
+            if not _jlink_libraries() and _src_jdk_version >= 9 and _src_jdk_version <= 14:
+                return '-J--add-modules=jdk.scripting.nashorn -J--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code=jdk.internal.vm.compiler'
```

Comment:
Instead of forcing `jdk.scripting.nashorn` onto the module path we should make sure that the substitutions that fail without are written in a way so that they can handle `jdk.scripting.nashorn` to be missing.

I.e. Define a new `BooleanSupplier` that is only true when module `jdk.scripting.nashorn` is available and add it to the `@TargetClass` annotations attribute value `onlyWith` e.g.
```java
@TargetClass(className = "jdk.nashorn.api.scripting.ClassFilter", onlyWith = {JDK14OrEarlier.class, hasModuleNashorn.class})
public final class Target_jdk_nashorn_api_scripting_ClassFilter {
}
```
and
```java
@TargetClass(className = "jdk.nashorn.api.scripting.NashornScriptEngineFactory", onlyWith = {JDK14OrEarlier.class, hasModuleNashorn.class})
public final class Target_jdk_nashorn_api_scripting_NashornScriptEngineFactory {
...
}
```

Reply:
OK, that seems a more elegant way of fixing it. Will fix.

Reply:
Fixed in the updated patch.